modules/howtos/examples/Import.java

```diff
@@ -0,0 +1,411 @@
+      });
+
+    Flux<MutationResult> results = 
+      rows
+      .map(row -> preprocess(row))
+      .flatMap(doc -> reactiveCollection.upsert(doc.getId(), doc.getContent()));
+      
+    results.subscribe(System.out::println);
+    results.blockLast(Duration.of(60, ChronoUnit.SECONDS));
+
+    System.out.println("DONE");
+    // end::omit[]
+  }
+// end::importTSV_batch[]
+  
+  // tag::importJSON[]
+  public void importJSON() {
+    try {
+      String content  = new String(
+        Files.readAllBytes(
+          Paths.get("howtos/examples/import.json")),
+        StandardCharsets.UTF_8);
+      
+      for (Object row: JsonArray.fromJson(content)) {
+        JsonObject json = ((JsonObject) row);
+        upsertRow(json.toMap());
+      }
+    }
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+  }
+  // end::importJSON[]
+  
+  // tag::importJSON_batch[]
+  public void importJSON_batch() {
+    // tag::omit[]
+    System.out.println("importJSON_batch");
+    // end::omit[]
+
+    try {
+      String content  = new String(
+        Files.readAllBytes(
+          Paths.get("howtos/examples/import.json")),
+        StandardCharsets.UTF_8);
+      
+      Flux<MutationResult> results = 
+        Flux.fromIterable(JsonArray.fromJson(content))
+          .map(row -> ((JsonObject) row).toMap())
+          .map(map -> preprocess(map))
+          .flatMap(doc -> reactiveCollection.upsert(doc.getId(), doc.getContent()));
+          
+      results.subscribe(System.out::println);
+      results.blockLast(Duration.of(60, ChronoUnit.SECONDS));
+    }
+    // ...
+    // tag::omit[]
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+    System.out.println("DONE");
+    // end::omit[]
+  }
+  // end::importJSON_batch[]
+  
+  // tag::importJSONL[]
+  public void importJSONL() {
+    try {
+        BufferedReader br =
+          new BufferedReader(
+            new FileReader("howtos/examples/import.jsonl"));
+        String line;
+        while ((line = br.readLine()) != null) {
+          Map<String,Object> row =
+            JsonObject.fromJson(line).toMap();
+            
+          upsertRow(row);
+      }
+    }
+    // ...
+    // tag::omit[]
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+    // end::omit[]
+  }
+  // end::importJSONL[]
+  
+  // tag::importJSONL_batch[]
+  public void importJSONL_batch() {
+    // tag::omit[]
+    System.out.println("importJSONL_batch");
+    // end::omit[]
+
+    Flux<String> lines = Flux.using(
```

Comment:
NIT: Here's how Simon recommends doing this (see the "conclusion" section of https://simonbasle.github.io/2017/10/file-reading-in-reactor/ )

    Flux<String> lines = Flux.defer(() ->
            Flux.fromStream(Files.lines(thePath)));

Reply:
Thanks - he uses the `.using` construct in the main body of the article though, and the `.defer` only in the footnoted caveat.

Also .using does some error handling which the .defer incantation doesn't so not sure we'd gain any clarity by switching here?